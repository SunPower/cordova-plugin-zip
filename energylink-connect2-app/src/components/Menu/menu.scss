@import '~@sunpower/theme-dark/styles/variables';

$headerHeight: 60px;

.menu-floating {
  position: absolute;
  top: $headerHeight;
  z-index: 30;
  height: calc(100% - #{$headerHeight});
  background-color: $black;
  width: 100%;
  margin-top: env(safe-area-inset-top);
  padding-bottom: calc(4 * env(safe-area-inset-bottom));
}

.menuitems {
  display: grid;
  grid-template-columns: 1fr 1fr;
  & > div > button {
    margin: auto;
  }
}
