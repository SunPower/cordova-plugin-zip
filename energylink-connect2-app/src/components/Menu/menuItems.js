import paths from 'routes/paths'
import GiveFeedback from 'pages/GiveFeedback'
import Firmwares from 'pages/Firmwares'
import VersionInformation from 'pages/VersionInformation'
import FirmwareReleaseNotes from 'pages/FirmwareReleaseNotes'

export default [
  {
    icon: 'sp-map',
    text: 'CHANGE_ADDRESS',
    to: paths.PROTECTED.ROOT.path,
    display: true
  },
  {
    icon: 'sp-update',
    text: 'MANAGE_FIRMWARES',
    component: Firmwares,
    display: true
  },
  {
    icon: 'sp-info',
    text: 'VERSION_INFORMATION',
    component: VersionInformation,
    display: true
  },
  {
    icon: 'sp-feedback',
    text: 'GIVE_FEEDBACK',
    component: GiveFeedback,
    display: true
  },
  {
    text: 'FIRMWARE_RELEASE_NOTES',
    component: FirmwareReleaseNotes,
    display: false
  }
]
