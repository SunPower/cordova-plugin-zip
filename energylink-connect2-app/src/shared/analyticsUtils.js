import {
  __,
  compose,
  curry,
  gt,
  isEmpty,
  join,
  lensIndex,
  map,
  not,
  over,
  props,
  propSatisfies,
  toLower,
  toUpper,
  when
} from 'ramda'

export const isNotEmpty = compose(not, isEmpty)

export const capitalizeWord = when(
  propSatisfies(gt(__, 0), 'length'),
  compose(join(''), over(lensIndex(0), toUpper), toLower)
)

const stringValueOrEmptyString = stringValue => stringValue || ''

const getStringProps = curry(compose(map(stringValueOrEmptyString), props))

export const getUserProfile = getStringProps([
  'uniqueId',
  'firstName',
  'lastName',
  'email',
  'dealerName',
  'recordType'
])
