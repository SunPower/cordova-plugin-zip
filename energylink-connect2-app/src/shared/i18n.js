import { useMemo } from 'react'
import { useSelector } from 'react-redux'

import * as languages from './locales'
import * as localizedComponents from './localizedComponents'

const getTranslate = translations => (key, ...params) => {
  let value = translations[key] || key
  if (params) {
    params.forEach((p, idx) => {
      value = value.replace(`{${idx}}`, p)
    })
  }
  return value
}

export const useI18n = () => {
  const language = useSelector(state => state.language)
  const { locale = 'en' } = language
  return useMemo(() => getTranslate(languages[locale]), [locale])
}

export const useI18nComponent = () => {
  const language = useSelector(state => state.language)
  const { locale = 'en' } = language || {}

  return useMemo(
    () => componentName => {
      const path = `${locale}_${componentName}`
      const Component = localizedComponents[path] || {}
      if (typeof Component !== 'function') {
        throw new Error(`Component not found: ${path}`)
      }
      return Component()
    },
    [locale]
  )
}
