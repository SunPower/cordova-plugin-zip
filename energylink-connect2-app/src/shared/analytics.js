import appVersion from '../macros/appVersion.macro'
import { capitalizeWord, getUserProfile } from 'shared/analyticsUtils'
import { MIXPANEL_EVENT_QUEUED } from 'state/actions/analytics'
import { getAppleDeviceFamily } from 'shared/appleDevicesTable'
import { prop } from 'ramda'

const getAppAndDeviceProperties = () => {
  const { device } = window
  const { model, version, platform, manufacturer } = device

  let deviceModel
  let deviceBrowserVersion

  if (platform === 'iOS') {
    deviceModel = getAppleDeviceFamily(model)
    deviceBrowserVersion = version
  } else {
    deviceModel = `${manufacturer}, ${model}`
  }

  return {
    'App Build': appVersion(),
    'App Flavor': process.env.REACT_APP_FLAVOR,
    'OS Version': version,
    ...(deviceBrowserVersion ? { $browser_version: deviceBrowserVersion } : {}),
    $device: deviceModel
  }
}

export const loggedIn = user => {
  const { mixpanel } = window
  const {
    uniqueId: userId,
    firstName,
    lastName,
    email,
    dealerName,
    recordType: dealerType
  } = getUserProfile(user)
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.identify(userId)
  mixpanel.people.set(appAndDeviceProperties)
  mixpanel.people.set({
    $first_name: firstName,
    $last_name: lastName,
    $email: email,
    'Dealer Name': dealerName,
    'Dealer Type': capitalizeWord(dealerType)
  })
  mixpanel.track('Login', { Success: true })

  return MIXPANEL_EVENT_QUEUED('Login - Success')
}

export const loginFailed = () => {
  const { mixpanel } = window
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.track('Login', { Success: false })

  return MIXPANEL_EVENT_QUEUED('Login - Failed')
}

export const scanPVS = ({ pvsSN, entryMethod }, event) => {
  const { mixpanel } = window

  mixpanel.register({ 'PVS SN': pvsSN })
  mixpanel.track('Scan PVS Tag', {
    'Entry Method': entryMethod
  })

  return MIXPANEL_EVENT_QUEUED(event)
}

export const siteNotFound = ({
  uniqueId: userId,
  firstName,
  lastName,
  email,
  dealerName,
  recordType: dealerType
}) => {
  const { mixpanel } = window
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.identify(userId)
  mixpanel.people.set(appAndDeviceProperties)
  mixpanel.people.set({
    $first_name: firstName,
    $last_name: lastName,
    $email: email,
    'Dealer Name': dealerName,
    'Dealer Type': capitalizeWord(dealerType)
  })
  mixpanel.track('Find Site', { Found: false })

  return MIXPANEL_EVENT_QUEUED('Find Site - site not found')
}

export const siteFound = (
  {
    uniqueId: userId,
    firstName,
    lastName,
    email,
    dealerName,
    recordType: dealerType
  },
  { city, st_id, postalCode, siteKey, commissioned }
) => {
  const { mixpanel } = window
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.identify(userId)
  mixpanel.people.set(appAndDeviceProperties)
  mixpanel.people.set({
    $first_name: firstName,
    $last_name: lastName,
    $email: email,
    'Dealer Name': dealerName,
    'Dealer Type': capitalizeWord(dealerType)
  })
  mixpanel.track('Find Site', {
    Found: true,
    ...{
      $city: city,
      State: st_id,
      'Zip Code': postalCode,
      'Site ID': siteKey,
      Commissioned: commissioned
    }
  })

  return MIXPANEL_EVENT_QUEUED('Find Site - site found')
}

export const saveConfiguration = config => {
  const { mixpanel } = window
  mixpanel.track('Configure', config)
  return MIXPANEL_EVENT_QUEUED('Configure - submit site config')
}

export const saveInventory = (inventory, hasEss) => {
  const { mixpanel } = window

  mixpanel.register({
    'Equinox Storage Site': hasEss
  })

  mixpanel.track('Inventory', {
    'AC Module Count': prop('AC_MODULES', inventory),
    'ESS System Size': hasEss ? prop('ESS', inventory) : 'None',
    'Equinox Storage Site': hasEss
  })
  return MIXPANEL_EVENT_QUEUED('Inventory')
}

export const deviceResume = pageResume => {
  const { mixpanel } = window
  mixpanel.register({
    Page: pageResume
  })
  return MIXPANEL_EVENT_QUEUED('User opens app - Register page resume')
}

/**
 *
 * @param {string}[interfaceType] - The type of connection the user has. either can be  PVS, Cell, or Customer WiFi
 * @return {Action<null, null>}
 */
export const internetConnection = (interfaceType = '') => {
  const { mixpanel } = window
  mixpanel.register({
    'Internet Connection': interfaceType
  })
  return MIXPANEL_EVENT_QUEUED('User opens app - Register page resume')
}

export const firmwareUpdateEvent = data => {
  const { mixpanel } = window
  mixpanel.track('Firmware Update', {
    'Current FW Version': prop('fromFWVersion', data),
    'Current BN Version': prop('fromBuildNumber', data),
    'New FW Version': prop('toFWVersion', data),
    'New BN Version': prop('toBuildNumber', data),
    Success: prop('success', data)
  })
  return MIXPANEL_EVENT_QUEUED('Firmware update - Update successful')
}

export const timeMixPanelEvent = name => {
  if (!name) throw new Error('The name parameter is required')
  const { mixpanel } = window
  mixpanel.time_event(name)
}

export const commissionSite = ({ duration, timeConfiguring }) => {
  const { mixpanel } = window
  mixpanel.track('Commission Site', {
    $duration: duration,
    'Time Configuring': timeConfiguring
  })
  return MIXPANEL_EVENT_QUEUED(
    'Commission Site - first successful site configuration'
  )
}
