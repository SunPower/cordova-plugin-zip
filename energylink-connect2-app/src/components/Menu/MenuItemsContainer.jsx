import React from 'react'
import { useDispatch } from 'react-redux'
import { useHistory } from 'react-router-dom'
import Tile from '@sunpower/tile'
import paths from 'routes/paths'
import { useI18n } from 'shared/i18n'
import { LOGOUT } from 'state/actions/auth'
import { MENU_DISPLAY_ITEM, MENU_HIDE } from 'state/actions/ui'
import { isDebug, either } from 'shared/utils'

function MenuItemsContainer({ items }) {
  const t = useI18n()
  const dispatch = useDispatch()
  const history = useHistory()

  const logout = () => {
    history.push(paths.PROTECTED.ROOT.path)
    dispatch(MENU_HIDE())
    dispatch(LOGOUT())
  }

  const onClickItem = item => () => {
    if (item.component) {
      dispatch(MENU_DISPLAY_ITEM(item.text))
    } else {
      dispatch(MENU_HIDE())
      history.push(item.to)
    }
  }

  const goToDebug = () => {
    dispatch(MENU_HIDE())
    history.push(paths.PROTECTED.DEBUG_PAGE.path)
  }

  return (
    <section className="is-flex tile is-vertical level full-height">
      <article className="is-flex space-around flow-wrap mt-15">
        {items.map(menuItem =>
          either(
            menuItem.display,
            <div className="mb-30" key={menuItem.text}>
              <Tile
                icon={menuItem.icon}
                text={t(menuItem.text)}
                onClick={onClickItem(menuItem)}
              />
            </div>
          )
        )}
      </article>

      {isDebug && (
        <button
          onClick={goToDebug}
          className="button is-primary is-outlined mt-10 mb-10"
        >
          {t('DEBUG_ROUTES')}
        </button>
      )}

      <button
        className="button has-text-centered is-uppercase mb-10 has-text-primary is-text"
        onClick={logout}
      >
        {t('BTN_LOG_OUT')}
      </button>
    </section>
  )
}

export default MenuItemsContainer
