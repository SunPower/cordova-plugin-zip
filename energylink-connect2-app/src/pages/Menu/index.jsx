import React from 'react'
import { useDispatch } from 'react-redux'
import Tile from '@sunpower/tile'
import paths from 'routes/paths'
import menuItems from 'components/Header/menuItems'
import { useI18n } from 'shared/i18n'
import { LOGOUT } from 'state/actions/auth'

function Menu(props) {
  const t = useI18n()
  const dispatch = useDispatch()

  const logout = () => {
    props.history.push(paths.PROTECTED.ROOT.path)
    dispatch(LOGOUT())
  }

  return (
    <section className="is-flex tile is-vertical level full-height">
      <article className="is-flex space-around flow-wrap mt-15">
        {menuItems.map(menuItem => (
          <div className="mb-30" key={menuItem.text}>
            <Tile
              icon={menuItem.icon}
              text={t(menuItem.text)}
              onClick={open(menuItem.to, props.history)}
            />
          </div>
        ))}
      </article>
      <button
        className="button has-text-centered is-uppercase mb-10 has-text-primary is-text"
        onClick={logout}
      >
        {t('BTN_LOG_OUT')}
      </button>
    </section>
  )
}

export default Menu

const open = (to, history) => () => history.push(to)
