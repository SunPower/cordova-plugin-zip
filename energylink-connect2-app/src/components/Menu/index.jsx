import clsx from 'clsx'
import { find, path, propEq, propOr } from 'ramda'
import React from 'react'
import { useSelector } from 'react-redux'

import menuItems from './menuItems'
import MenuItemsContainer from './MenuItemsContainer'

import { either } from 'shared/utils'
import './menu.scss'

function Menu() {
  const menuItemToDisplay = useSelector(path(['ui', 'menu', 'itemToDisplay']))
  const show = useSelector(path(['ui', 'menu', 'show']))

  const selectedMenuItem = find(propEq('text', menuItemToDisplay))(menuItems)
  const Component = propOr(null, 'component', selectedMenuItem)

  return (
    <div
      className={clsx('menu-floating', {
        'is-hidden': !show
      })}
    >
      {either(!selectedMenuItem, <MenuItemsContainer items={menuItems} />)}
      {selectedMenuItem && <Component />}
    </div>
  )
}

export default Menu
