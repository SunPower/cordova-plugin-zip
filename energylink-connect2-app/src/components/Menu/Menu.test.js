import { shallow } from 'enzyme'
import React from 'react'
import * as ReactDOM from 'react-dom'
import * as reactRedux from 'react-redux'

import Menu from '.'

import * as i18n from 'shared/i18n'

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    history: {
      push: jest.fn()
    }
  })
}))

describe('Menu Page', () => {
  let dispatchMock

  beforeEach(() => {
    dispatchMock = jest.fn()
    ReactDOM.createPortal = jest.fn((element, node) => {
      return element
    })

    jest.spyOn(reactRedux, 'useDispatch').mockImplementation(() => dispatchMock)
    jest.spyOn(reactRedux, 'useSelector').mockImplementation(jest.fn)
    jest
      .spyOn(i18n, 'useI18n')
      .mockImplementation(path => (key, ...params) =>
        `${key.toUpperCase()} ${params.join('_')}`.trim()
      )
  })

  test('render correctly', () => {
    const component = shallow(<Menu />)
    expect(component).toMatchSnapshot()
  })
})
