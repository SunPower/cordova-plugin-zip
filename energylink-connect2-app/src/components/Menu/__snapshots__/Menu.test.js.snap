// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Page render correctly 1`] = `
<div
  className="menu-floating"
>
  <MenuItemsContainer
    items={
      Array [
        Object {
          "display": true,
          "icon": "sp-map",
          "text": "CHANGE_ADDRESS",
          "to": "/",
        },
        Object {
          "component": [Function],
          "display": true,
          "icon": "sp-update",
          "text": "MANAGE_FIRMWARES",
        },
        Object {
          "component": [Function],
          "display": true,
          "icon": "sp-info",
          "text": "VERSION_INFORMATION",
        },
        Object {
          "component": [Function],
          "display": true,
          "icon": "sp-feedback",
          "text": "GIVE_FEEDBACK",
        },
        Object {
          "component": [Function],
          "display": false,
          "text": "FIRMWARE_RELEASE_NOTES",
        },
      ]
    }
  />
</div>
`;
